src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  let newState = {};
+  const historyState = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        newState = { ...copyState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        newState = { ...copyState };
+        action.keysToRemove.forEach((key) => delete newState[key]);
+        break;
+      case 'clear':
+        newState = {};
+        break;
+      default:
+        newState = copyState;
+        break;
+    }
+
+    historyState.push(newState);
+    copyState = newState;
```

Comment:
This line is mutating `copyState` which is not necessary. You should work with `newState` directly in each iteration and then push a copy to `historyState`.